cpp/src/arrow/ipc/writer.cc

```diff
@@ -1147,13 +1201,16 @@ Status SerializeSchema(const Schema& schema, MemoryPool* pool,
   std::shared_ptr<io::BufferOutputStream> stream;
   RETURN_NOT_OK(io::BufferOutputStream::Create(1024, pool, &stream));
 
-  DictionaryMemo memo;
-  SchemaWriter schema_writer(schema, &memo, pool, stream.get());
+  // XXX: PayloadStreamWriter takes a std::shared_ptr<Schema> but we only got a
+  // raw reference.  We fake it with a null deleter.
+  auto deleter = [](Schema*) -> void {};
+  std::shared_ptr<Schema> schema_ptr(const_cast<Schema*>(&schema), deleter);
 
-  // Unused
-  std::vector<FileBlock> dictionary_blocks;
+  auto payload_writer = make_unique<PayloadStreamWriter>(stream.get());
+  RecordBatchPayloadWriter writer(std::move(payload_writer), schema_ptr);
```

Comment:
`RecordBatchStreamWriter` takes `io::OutputStream*` as its first argument.
If `writer` should own `payload_writer`, then the parameter list should be modified to take `unique_ptr<io::OutputStream>`.
If `writer` doesn't own `payload_writer`:
```suggestion
  RecordBatchPayloadWriter writer(payload_writer.get(), schema_ptr);
```

Reply:
This is `RecordBatchPayloadWriter`, not `RecordBatchStreamWriter`.